src/base/bittorrent/session.cpp

```diff
@@ -2368,6 +2368,36 @@ void Session::setIPFilterFile(QString path)
     }
 }
 
+void Session::setBannedIPs(const QStringList &newList)
+{
+    if (newList == m_bannedIPs)
+        return; // do nothing
+    // malformed IP strings will be silently filtered out
+    QStringList filteredList;
+    foreach (const QString &ip, newList) {
+        boost::system::error_code ec;
+        libt::address::from_string(ip.toLatin1().constData(), ec); // only error code needed
+        if (!ec) {
+            filteredList << ip;
+        }
+        else {
+            Logger::instance()->addMessage(
+                        tr("%1' is not a valid IP address and was rejected while applying the list of banned addresses.")
+                        .arg(ip)
+                        , Log::INFO);
+        }
+    }
+    // store to session settings
+    m_bannedIPs = filteredList;
+    m_IPFilteringChanged = true;
+    configureDeferred();
```

Comment:
Unless I am mistaken the above 2 lines are useless. All they do is re-parse the ip_filter file and apply it. The new `bannedIPs` aren't applied.
This is done in `Session::processBannedIPs()` which isn't called. But it isn't enough to call it here and be done with it. There is subtle bug in this whole logic. Consider the following:
1. The logic of `Session::processBannedIPs()` is to ask **libtorrent** for the current filter and then append to it the bannedIPs.
2. User has banned 3 IPs.
3. Edits the list, removes one and adds two more.
4. Due to the logic of `Session::processBannedIPs()` we won't unban the deleted IP.

I have a proposal on how to resolve this:
1. If `newList` and `m_bannedIPs` are different, then find the difference(what was deleted from m_bannedIPs).
2. Extend `Session::processBannedIPs()` to take a QStringList argument with a default value of an empty list.
3. Then inside there, if the list isn't empty apply the rules for that list with `0` instead of `libt::ip_filter::blocked` as a flag.